jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/vsdm/ReqResForwarder.java

```diff
@@ -0,0 +1,147 @@
+import us.dot.its.jpo.ode.j2735.J2735;
+import us.dot.its.jpo.ode.j2735.semi.ConnectionPoint;
+import us.dot.its.jpo.ode.j2735.semi.IPv4Address;
+import us.dot.its.jpo.ode.j2735.semi.IpAddress;
+import us.dot.its.jpo.ode.j2735.semi.PortNumber;
+import us.dot.its.jpo.ode.j2735.semi.ServiceRequest;
+import us.dot.its.jpo.ode.j2735.semi.ServiceResponse;
+
+/*
+ * This class receives service request from the OBU and forwards it to the SDC.
+ * It also receives service response from SDC and forwards it back to the OBU.
+ */
+public class ReqResForwarder implements Runnable {
+	private OdeProperties odeProps;
+	private static Coder coder = J2735.getPERUnalignedCoder();
+	private Logger logger = LoggerFactory.getLogger(this.getClass());
+	private DatagramSocket socket = null;
+	private byte[] payload;
+	private String obuReturnAddr;
+	private int obuReturnPort;
+
+	public ReqResForwarder(OdeProperties odeProps, ServiceRequest request, String obuIp, int obuPort) {
+		this.odeProps = odeProps;
+		this.obuReturnAddr = obuIp;
+		this.obuReturnPort = obuPort;
+		this.payload = createRequest(request);
+		try {
+			socket = new DatagramSocket(odeProps.getForwarderPort());
+			logger.info("ODE: Created depositor Socket with port " + odeProps.getForwarderPort());
+		} catch (SocketException e) {
+			logger.error("ODE: Error creating socket with port " + odeProps.getForwarderPort(), e);
+		}
+	}
+
+	public byte[] createRequest(ServiceRequest request) {
+		IpAddress ipAddr = new IpAddress();
+		ipAddr.setIpv4Address(new IPv4Address(J2735Util.ipToBytes(odeProps.getReturnIp())));
+		ConnectionPoint newReturnAddr = new ConnectionPoint(ipAddr, new PortNumber(odeProps.getReturnPort()));
+
+		if (request.hasDestination()) {
+			logger.info("Received Service Request contains destination field");
+			logger.info("Old OBU destination IP: {} Source Port: {}", this.obuReturnAddr, this.obuReturnPort);
+			byte[] ipBytes = request.getDestination().getAddress().getIpv4Address().byteArrayValue();
+			this.obuReturnAddr = J2735Util.ipToString(ipBytes);
+			this.obuReturnPort = request.getDestination().getPort().intValue();
+			logger.info("New OBU destination IP: {} Source Port: {}", this.obuReturnAddr, this.obuReturnPort);
+		}
+
+		request.setDestination(newReturnAddr);
+		logger.info("New ODE destination IP: {} Source Port: {}", odeProps.getReturnIp(), odeProps.getReturnPort());
+
+		ByteArrayOutputStream sink = new ByteArrayOutputStream();
+		try {
+			coder.encode(request, sink);
+		} catch (EncodeFailedException | EncodeNotSupportedException e) {
+			logger.error("ODE: Error Encoding VSD Deposit ServiceRequest", e);
+		}
+
+		return sink.toByteArray();
+	}
+
+	public void send() {
+		try {
+			logger.info("\nODE: Printing ServiceRequest in hex: \n{}\n", Hex.encodeHexString(payload));
+			logger.info("ODE: Sending VSD Deposit ServiceRequest to IP: {} Port: {}", odeProps.getSdcIp(),
+					odeProps.getSdcPort());
+			socket.send(new DatagramPacket(payload, payload.length,
```

Comment:
As per Leidos recommendation, you need to launch a separate threat (from a thread pool) to receive the response before sending the request in order to avoid the risk of missing the response. In fact, this thread could be just that, the thread that receives the response. The sending of the request can be done by the same thread that will be depositing the VSDs.